plugins/obs-ffmpeg/texture-amf.cpp

```diff
@@ -0,0 +1,1129 @@
+{
+	constexpr int64_t amf_timebase = AMF_SECOND;
+	return ts * amf_timebase / (int64_t)enc->fps_den;
+}
+
+static inline int64_t convert_to_obs_ts(amf_data *enc, int64_t ts)
+{
+	constexpr int64_t amf_timebase = AMF_SECOND;
+	return ts * (int64_t)enc->fps_den / amf_timebase;
+}
+
+static void convert_to_encoder_packet(amf_data *enc, AMFDataPtr &data,
+				      encoder_packet *packet)
+{
+	if (!data)
+		return;
+
+	enc->packet_data = AMFBufferPtr(data);
+	data->GetProperty(L"PTS", &packet->pts);
+
+	bool hevc = enc->codec == amf_codec_type::HEVC;
+	const wchar_t *get_output_type =
+		hevc ? AMF_VIDEO_ENCODER_HEVC_OUTPUT_DATA_TYPE
+		     : AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE;
+
+	uint64_t type;
+	data->GetProperty(get_output_type, &type);
+
+	switch (type) {
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_IDR:
+		packet->priority = OBS_NAL_PRIORITY_HIGHEST;
+		break;
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_I:
+		packet->priority = OBS_NAL_PRIORITY_HIGH;
+		break;
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_P:
+		packet->priority = OBS_NAL_PRIORITY_LOW;
+		break;
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_B:
+		packet->priority = OBS_NAL_PRIORITY_DISPOSABLE;
+		break;
+	}
+
+	packet->data = (uint8_t *)enc->packet_data->GetNative();
+	packet->size = enc->packet_data->GetSize();
+	packet->type = OBS_ENCODER_VIDEO;
+	packet->dts = convert_to_obs_ts(enc, data->GetPts());
+	packet->keyframe = type == AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_IDR;
+}
+
+static bool amf_encode_tex(void *data, uint32_t handle, int64_t pts,
+			   uint64_t lock_key, uint64_t *next_key,
+			   encoder_packet *packet, bool *received_packet)
+try {
+	amf_data *enc = (amf_data *)data;
+	ID3D11DeviceContext *context = enc->context;
+	ComPtr<ID3D11Texture2D> output_tex;
+	ComPtr<ID3D11Texture2D> input_tex;
+	ComPtr<IDXGIKeyedMutex> km;
+	AMFSurfacePtr amf_surf;
+	AMFDataPtr amf_out;
+	AMF_RESULT res;
+
+	if (handle == GS_INVALID_HANDLE) {
+		*next_key = lock_key;
+		throw "Encode failed: bad texture handle";
+	}
+
+	get_tex_from_handle(enc, handle, &km, &input_tex);
+	if (!enc->texture)
+		create_texture(enc, input_tex);
+	output_tex = enc->texture;
+
+	/* ------------------------------------ */
+	/* copy to output tex                   */
+
+	km->AcquireSync(lock_key, INFINITE);
+	context->CopyResource((ID3D11Resource *)output_tex.Get(),
+			      (ID3D11Resource *)input_tex.Get());
+	context->Flush();
+	km->ReleaseSync(*next_key);
+
+	/* ------------------------------------ */
+	/* map output tex to amf surface        */
+
+	/* XXX: does this really need to be recreated every frame? */
+
+	res = enc->amf_context->CreateSurfaceFromDX11Native(output_tex,
+							    &amf_surf, nullptr);
+	if (res != AMF_OK)
+		throw amf_error("CreateSurfaceFromDX11Native failed", res);
+
+	int64_t last_ts = convert_to_amf_ts(enc, pts - 1);
+	int64_t cur_ts = convert_to_amf_ts(enc, pts);
+
+	amf_surf->SetPts(cur_ts);
+	amf_surf->SetProperty(L"PTS", pts);
+	amf_surf->SetDuration(cur_ts - last_ts);
```

Comment:
There is no need to SetPts and SetDuration https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/275 .

I see no need to convert_to_amf_ts either. The only reason to SetProperty("PTS") is so that we can later retrieve and set this for the encoder out packet. That value seems like it should be the same as what obs gave as the input pts.

I'm doing this here https://github.com/e00E/obs-amf/blob/05986543636d8e1a795153de3e4f52087197cf6a/source/encoder.cpp#L270 .

Reply:
I don't know about the duration, but the SetPts() seems definitely required because it provides the DTS on the resulting packet. See my commentary above.